### PR TITLE
Better utilizing GADT bounds for HKTs

### LINKED ISSUE #13257: Fail to utilize HKT GADT bounds for applied types
## Compiler version

3.0.1

## Minimized code

```Scala
type Const = [X] =>> Int

trait Expr[F[_]]
case class ConstExpr() extends Expr[Const]

def foo[F[_], A](e: Expr[F]): F[A] = e match
  case _: ConstExpr => 0
```

## Output

```scala
Found:    (0 : Int)
Required: F[A]

where:    F is a type in method foo which is an alias of [X] =>> Int
```

## Expectation

By pattern matching we derive the GADT bound `F  := [X] =>> Int` (as shown in the error message), but during type comparison the compiler will not recognize `F[X] == Int` for the applied type.

### PR DESCRIPTION
**Pull request by Linyxus**
This PR fixes #13257. Before this PR, the following code example could not compile:
```scala
type Const = [X] =>> Int

trait Expr[+F[_]]
case class ConstExpr() extends Expr[Const]

def foo[F[_], A](e: Expr[F]): F[A] = e match
  case _: ConstExpr => 0
```

It is sound to allow its compilation, since `Int <: Const[A] <: F[A]`. This PR changes the TypeComparer's behavior by trying to approximate the type constructor when GADT bounds is available. A global state is added to the TypeComparer to track whether HKT bounds have been used during type comparison, to properly handle the injectivity conditions.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala [hunk 1]
```diff
@@ -1092,6 +1093,7 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
                     val tyconIsInjective =
                       (tycon1sym.isClass || tycon2sym.isClass)
                       && (!touchedGADTs || gadtIsInstantiated)
+                      && !frozenGadt
```
Comment (Linyxus) [inline compiler/src/dotty/tools/dotc/core/TypeComparer.scala:1095]: Note that this is a workaround for the frozenGadt problem fixed by #13380. So maybe we should first merge #13380, remove this line and see if any test breaks, and then merge this PR.

**Comment by Linyxus** [comment]
@abgruszecki Hey Alex, I have finished the modification. Also, I add some additional code to allow the LHS HKT to be approximated to its GADT upper bound.

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala [hunk 2]
```diff
@@ -1213,11 +1225,21 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
...
+              }
+            } && { GADTused = true; true }
+
+          {
+            !sym.isClass && {
+              defn.isCompiletimeAppliedType(sym) && compareCompiletimeAppliedType(tp1, tp2, fromBelow = false) ||
+              recur(tp1.superType, tp2) ||
+              tryLiftedToThis1
+            }
+          } || byGadtBounds
```
Comment (abgruszecki) [inline compiler/src/dotty/tools/dotc/core/TypeComparer.scala:1242]: The outer braces aren't necessary here.

**Review by abgruszecki — APPROVED** [review]
Aside from the nitpick, the PR LGTM.

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala (hunk 2 above)
Comment (Linyxus) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala]: Whoops. Thank you for point that out! Fixed. ;)

**Pull request merged**